nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayStreamSupportPositiveCases.java

```diff
@@ -191,7 +191,8 @@ private Stream<T> test1(Stream<T> stream) {
 
     private Stream<T> test2(Stream<T> stream) {
       Preconditions.checkNotNull(ref);
-      // BUG: Diagnostic contains: dereferenced expression ref is @Nullable
+      // no error since we propagate nullability facts to stream callbacks, which
+      // in sane code are invoked soon after the stream is created
```

Comment:
I guess I go a bit back and forth between "this is probably safe" and "restricting it to cases where the stream doesn't escape the method will remove most false positives too and be safer". Your call.

Reply:
Let me think through a bit more what it would take to implement the safer version.

Reply:
@lazaroclapp I thought about this more and I think a reasonably precise safer version would be kind of a pain to implement.  My thought was to try to detect if the overall chain of stream calls ended in a terminal operation.  But there are a bunch of terminal operations on `Stream` itself, and then there are methods to convert `Stream` to `IntStream`, `DoubleStream`, etc., each with their own terminal operations.  (As an aside, in a follow-up we should probably handle those primitive stream types in the same way that we handle `Stream`s themselves.)  We could also just check if the enclosing method has a return type of `Stream`, but that is imprecise, and also misses cases like passing as a parameter or storing in a field.

Bottom line, my feeling is at this point, we are most likely ok going with the less safe version.  But, I think it's time we add a wiki page on our stream handling in general, documenting how we propagate nullability facts from `filter()` methods to subsequent callbacks and also this new propagation of nullability facts from the enclosing context.  On that page we could give caveats on cases where our handling could be unsound, and why we decided to do what we did.

Let me know what you think.

Reply:
Of course, what this really needs is a new type system about the state of `Stream` objects... but in the absence of that, just eagerly propagating nullability of APs in the closure is probably safe enough (kidding about the `Stream` type-state system... mostly)

Reply:
To close the loop here I added a wiki page with some docs:

https://github.com/uber/NullAway/wiki/Stream-Handling